lnst/RecipeCommon/endpoints.py

```diff
@@ -0,0 +1,87 @@
+from __future__ import annotations
+from dataclasses import dataclass
+from typing import Generic, Iterator, Protocol, TypeVar
+
+from lnst.Common.IpAddress import BaseIpAddress
+from lnst.Controller.Namespace import Namespace
+from lnst.Devices.Device import Device
+
+
+class EndpointLike(Protocol):
+    @property
+    def host(self) -> Namespace:
+        ...
+
+    @property
+    def device_name(self) -> str:
+        ...
+
+
+TIpAddress_co = TypeVar("TIpAddress_co", bound=BaseIpAddress, covariant=True)
```

Comment:
the typing in this module could probably use some docstring explanations or examples so that it's easier to understand what usecases are expected and attempted to be represented with the types.

Reply:
Added docstrings to the classes. Let me know if they make more sense now.

Reply:
Do you want me to explain what covariant means? Or why did I use it maybe? I used it because I want to be able to assign a value of type `EndpointPair[IPEndpoint]` to a variable of type `EndpointPair[Endpoint]`. If the typevar wasn't covariant, this wouldn't work.

Should I put this in a comment there?

Reply:
>  I used it because I want to be able to assign a value of type EndpointPair[IPEndpoint] to a variable of type EndpointPair[Endpoint]. If the typevar wasn't covariant, this wouldn't work.

i think this should be the comment in the code then.

Reply:
Added the comment there.